bpo-17611: Move unwinding of stack from interpreter to compiler

nascheme opened the pull request:
This is a re-base and slightly more polished version of PR #2827.  I think this change is a good idea as it reduces the size of the stacked used by Python frames.  

Also, it makes the bytecode have an "ahead of time" computable effect on the stack.  See the changes in PyCompile_OpcodeStackEffect().  E.g. before WITH_CLEANUP_FINISH had a "XXX" comment saying it dropped at least one stack item but at runtime maybe more.  After this change, WITH_CLEANUP_FINISH always drops 6 items. 

I understand that is important for tools that try to JIT compile Python bytecode.  Having an ahead of time computable stack effect makes the JIT a lot simpler.

<!-- issue-number: bpo-17611 -->
https://bugs.python.org/issue17611
<!-- /issue-number -->

pitrou committed:
bpo-17611: Move unwinding of stack from interpreter to compiler

pitrou committed:
Cleanup

pitrou committed:
Improve stack size calculation for SETUP_{EXCEPT,FINALLY}

pitrou committed:
Remove useless code

pitrou committed:
Don't predict FOR_ITER as it breaks tracing without computed gotos

pitrou committed:
Fix test_dis

pitrou committed:
Add JUMP_FINALLY to avoid finally block duplication

pitrou committed:
Remove finally block duplication

pitrou committed:
More precise computation of code object stack size

pitrou committed:
JUMP_FINALLY pushes 6 values to be consistent with the effect of an exception

pitrou committed:
Exact stack size computation by popping stale exception state in RERAISE

pitrou committed:
Remove the now pointless POP_MANY

pitrou committed:
Add comment for RERAISE

pitrou committed:
Remove JUMP_FINALLY

pitrou committed:
Remove last block duplication

pitrou committed:
Get rid of END_ITER

pitrou committed:
Fix comments in frameobject.c

pitrou committed:
Add stack size tests

serhiy-storchaka committed:
Simplify handling of 'for' loops in frame_setlineno() and fix it for long opargs.

pitrou committed:
Update frozen importlib

pitrou committed:
Remove dead code and fix tracing tests

nascheme committed:
Fix tests in test_dis to match new bytecode.

nascheme committed:
Slightly more logical order for fblocktype enum.

nascheme committed:
Add news file.

ncoghlan commented on Lib/test/test_peepholer.py:
```diff
@@ -261,7 +261,7 @@ def f(cond1, cond2):
         returns = [instr for instr in dis.get_instructions(f)
                           if instr.opname == 'RETURN_VALUE']
-        self.assertEqual(len(returns), 6)
+        self.assertLessEqual(len(returns), 6)
```
Is this truly non-deterministic now? Or should the expected number of return opcodes be modified instead of making the check more permissive?

nascheme replied:
I looks to me like that check got changed in an old version of the patch and maybe isn't needed.  I'm going to review all the unit test changes and see if they are still necessary with the final version of the patch.  Antoine went through quite a few revisions.

nascheme replied:
I believe this part of the patch can be reverted.  self.assertEqual(len(returns), 6) works and I see no reason why the number of return statements should be indeterminate.

nascheme replied:
Sorry, the change below, "len(returns), 2", is the one that can be reverted. This one now returns 5.  I believe the reason is that there is no longer a SETUP_LOOP opcode and the peephole op is now eliminating (correctly, as far as I see), the "return 6" at the end of the function body.  So, should always be 5 return statements in bytecode now.

ncoghlan commented on Lib/test/test_peepholer.py:
```diff
@@ -275,7 +275,7 @@ def f(cond1, cond2):
         returns = [instr for instr in dis.get_instructions(f)
                           if instr.opname == 'RETURN_VALUE']
-        self.assertEqual(len(returns), 2)
+        self.assertLessEqual(len(returns), 2)
```
As for previous comment - can we adjust the number of expected returns here instead of relaxing the check?

ncoghlan commented on Python/ceval.c:
```diff
@@ -1869,67 +1876,67 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
+            PyObject *tb = POP();
+            if (exc == NULL) {
+                int i;
+                assert(val == NULL);
```
These should be regular validity checks, rather than assertions - bytecode editing may inject a `RERAISE` at a point where there isn't a valid exception state on top of the stack, and that should give a Python exception rather than a C level assertion failure.

nascheme replied:
I'm surprised to hear that.  Do we really expect ceval to be resistant to crashing if fed arbitrary bytecode?  That seems a tough standard to meet and would seem to imply a lot of overhead in checking that the bytecode doesn't do anything unsafe (e.g. re whats on the stack). If that is the case, I expect this patch needs a lot of fixing yet.

ncoghlan replied:
We don't always check directly in ceval - we'll often rely on the fact that ceval calls public CPython C APIs, and those already have their own error checking.

But we do have a lot of checks for "impossible" situations (search for SystemError in ceval.c), and the general principle of "bytecode hacking may cause an exception, but it won't crash the process" applies. (That said, there are some existing `assert` calls related to the stack state for function calls, so this could be worth raising on python-dev)

nascheme replied:
What kinds of bytecode changes are allowed?  I wrote a very simple minded fuzzer for ceval, giving it randomly generated code.  It crashes Python 3.6 immediately.  I don't disagree with better checking in principle but it seems a very high standard to meet without having some significant performance affects.

ncoghlan replied:
Yeah, given the existing assert() checks for function calls, I'll withdraw this comment for now.

ncoghlan commented on Python/ceval.c:
```diff
@@ -1869,67 +1876,67 @@ _PyEval_EvalFrameDefault(PyFrameObject *f, int throwflag)
-                Py_DECREF(status);
-                goto error;
+            else {
+                assert(PyExceptionClass_Check(exc));
```
As above - this should be a check that raises `PyExc_SystemError`, not a C level assertion.

ncoghlan reviewed (dismissed):
Reviewing the eval loop changes, my main comment is that we can't safely make assertions about the expected eval stack state when a particular bytecode is executed, since synthetically constructed bytecode may fail to meet those expectations in weird and exciting ways.

Instead, we should use explicit expectation checks, and raise `SystemError` when they're not met.

(Note: I haven't reviewed the compiler changes yet)

nascheme committed:
Revert this change to test_peepholer.py.

Behavior is still the same after the bytecode and compiler
changes.

nascheme committed:
Change test_peepholer check back to assertEqual().

The bytecode changes cause the number of expected returns
to go from 6 to 5.  The last return in the function now gets
eliminated, as it is unreachable.  I believe the SETUP_LOOP
opcode was defeating the peephole optimizer previously.  That's just
a guess though, I don't understand how the peephole optimizer works.

nascheme committed:
Pass 'final' block to compiler_push_finally_try().

The argument is not actually used for anything but it makes
more sense to pass 'final' rather than passing 'body'.

nascheme committed:
Bugfix for fblock_unwind_finally_try(), don't unwind current block.

If the finally body contains an exit (break or return), don't unwind
the finally body a second time.  That would cause an infinite loop
in the compiler.  We temporarily pop the top block, emit code for
the finally clause and then push the block back on.

At some (hopefully correct) comments explaining what's going on.

nascheme committed:
Make fblock stack into a singly linked list.

nascheme committed:
Introduce POP_NO_EXCEPT opcode.

When there is a 'return' inside the finally body of a try/finally,
we need to clear the current exception.  If the final body was
entered due to an exception, we also need to pop the EXCEPT_HANDLER
fbock.

pitrou commented on Lib/test/test_compile.py:
```diff
@@ -709,5 +709,102 @@ def test_func_and(self):
+            """
+        self.check_stack_size(snippet)
+
+    # This one unfortunately "leaks" a few stack slots for each snippet
```
Do you have any idea how to solve this one?

nascheme committed:
Make PUSH_NO_EXCEPT push a single NULL.

This is slightly more efficient.  For computing the stack effect
of opcodes, consider PUSH_NO_EXCEPT to use six stack slots so that
the stack will be large enough to handle an exception.

Update dis.rst (with content from PR 5006 by Serhiy).

nascheme committed:
Restore SETUP_WITH and SETUP_ASYNC_WITH.

Fixes problem pointed out by Serhiy:

There is a gap between calling the __enter__ method and the
SETUP_FINALLY instruction. If the exception is raised in the gap,
the __exit__ method will be never called. For example:

    a = []
    with CM() as a[0]: # IndexError
        ...

nascheme committed:
Refactor test_dis to make regen of expected instructions cleaner.

nascheme committed:
Fix test_dis.
